pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -379,6 +382,16 @@ public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentit
       return new BrokerResponseNative(QueryException.getException(QueryException.BROKER_TIMEOUT_ERROR, errorMessage));
     }
 
+    // Validate the request
+    try {
+      validateNumSegments(offlineRoutingTable, realtimeRoutingTable, _querySegmentLimit);
```

Comment:
I suggest using a negative value as default, but yes. Please retain original behavior when it is not configured.

Also, why do we want to skip at the query processing stage? Why not stop ingestion if the number of segments exceeds some "target" value (set on a per table basis).

Reply:
Because this is post-prune. So if your post-pruned segments go over a limit, it cuts you off. Once we have time based pruning, this becomes very powerful.

Reply:
Maybe I am missing the context, but are you trying to avoid cut-off because the queries become slower? If so, a better way is to honor the max time limit (which, I believe, is now being passed to the server via PinotQuery request). Is this a part of lazy load, or does this belong in a different PR? If it is a part of lazy load, can you share some information of how this is related? You can do it in the design doc. Thanks

Reply:
This is part of lazy loading. It's a guard to keep you from causing tons of segments to be pulled from the deep store to a local disc that may not have very much space, since it's meant more for random access of a few segments; not pulling a whole table.